cpp/src/arrow/compute/kernels/hash_aggregate.cc

```diff
@@ -2758,6 +2758,317 @@ struct GroupedOneFactory {
   InputType argument_type;
 };
 
+// ----------------------------------------------------------------------
+// List implementation
+
+template <typename Type, typename Enable = void>
+struct GroupedListImpl final : public GroupedAggregator {
+  using CType = typename TypeTraits<Type>::CType;
+  using GetSet = GroupedValueTraits<Type>;
+
+  Status Init(ExecContext* ctx, const std::vector<ValueDescr>&,
+              const FunctionOptions* options) override {
+    ctx_ = ctx;
+    // out_type_ initialized by GroupedListInit
+    values_ = TypedBufferBuilder<CType>(ctx_->memory_pool());
+    groups_ = TypedBufferBuilder<uint32_t>(ctx_->memory_pool());
+    values_bitmap_ = TypedBufferBuilder<bool>(ctx_->memory_pool());
+    return Status::OK();
+  }
+
+  Status Resize(int64_t new_num_groups) override {
+    num_groups_ = new_num_groups;
+    return Status::OK();
+  }
+
+  Status Consume(const ExecBatch& batch) override {
+    const auto* groups = batch[1].array()->GetValues<uint32_t>(1);
```

Comment:
I would maybe `DCHECK_EQ(batch[0].array()->offset, 0);` just to be safe (and make explicit the assumption here)

Reply:
Is this somehow related to this?
For columns 1 and 4 (zero indexed) in the table above for `use_exec_plan = true` I get:
```cpp
// col 1 - boolean
[
  true,
  true,
  false,
  false,
  null,
  null,
  false,
  false,
  true,
  false
]

// col 4 fixed size binary
[
  null,
  616161,
  626163,
  null,
  null,
  null,
  null,
  626162,
  313233,
  null
]
```

While it appears to be correct with `use_exec_plan = false`:
```cpp
// col 1 - boolean
[
  true,
  true,
  false,
  false,
  null,
  true,
  false,
  false,
  true,
  true
]

// col 4 - fixed sized binary
[
  null,
  616161,
  626163,
  null,
  323334,
  646464,
  626364,
  626162,
  313233,
  null
]
```

Am I handling these type incorrectly in the implementation? (These are without the `DCHECK_EQ`)
What am I missing?

Reply:
Also, boolean type also has problems with `use_threads = true`:
```cpp
[
  true,
  true,
  true,
  false,
  null,
  false,
  false,
  true,
  false,
  false
]
```

Reply:
Well then the assumption is incorrect, and we need to handle offsets.

Reply:
GetValues<CType> naively tries to adjust the returned pointer to account for the offset but it is unable to do this for boolean type. So we should use `GetValues<CType>(1, 0)` to prevent it from adjusting the offset, and then our code needs to handle the offset itself (e.g. AppendBuffers needs to take an offset)